riscv: Add support for floating point

xodus7 opened the pull request:
This change adds full shared floating point support for the RISCV
architecture with minimal impact of threads with floating point
support not enabled.

Signed-off-by: Corey Wharton <coreyw7@fb.com>

Commit:
riscv: Add support for floating point

This change adds full shared floating point support for the RISCV
architecture with minimal impact on threads with floating point
support not enabled.

Signed-off-by: Corey Wharton <coreyw7@fb.com>

xodus7 commented:
This change is in part derived for the the work done in PR #19095 which seems to have been abandoned. It's contributions on top of that change are:

* A few serious bug fixes in the original PR, including an issue that prevents the fp_sharing/generic test from ever passing.
* General code clean-up
* Ensures that the FPU is enabled/disabled properly in a per thread basis in mstatus register.
* Implements arch_float_disable() & arch_float_enable()
* Only saves FP context for threads with K_FP_REGS option to minimize impact of normal threads.
* Extend fp_sharing tests for RISCV including some special test cases in fp_sharing/general

This patch was fully tested under the qemu_riscv32 (hard and soft float, 32-bit & 64-bit FLEN) and qemu_riscv64 (hard-float, 64-bit FLEN). However, I had to create a patch in order to enable F & D extensions for the sifive_e machine. I'll be submitting this patch to the QEMU project shortly, but that means I cannot enable emulated testing for this change just yet. Once the required support to QEMU is added I can submit my patch to enabled the required tests.

xodus7 commented:
The relevant QEMU patchset: http://patchwork.ozlabs.org/project/qemu-devel/list/?series=164096

nategraff-sifive reviewed (approved):
This LGTM. Great work, thank you so much!

carlescufi commented:
@andrewboie @kgugala could you please take a look?

ioannisg commented on arch/Kconfig:
```diff
@@ -509,7 +509,7 @@ config MPU_GAP_FILLING
 	bool "Floating point"
 	depends on CPU_HAS_FPU
-	depends on ARM || X86 || ARC
+	depends on ARM || X86 || ARC || RISCV
```
Could this boolean sentence be sorted alphabetically?

ioannisg commented on tests/kernel/fp_sharing/generic/src/main.c:
```diff
@@ -233,7 +239,7 @@ void load_store_low(void)
 		 * disable floating point operations for the task.
 		 */
-#if (defined(CONFIG_X86) && defined(CONFIG_LAZY_FP_SHARING)) || \
+#if (defined(CONFIG_X86) && defined(CONFIG_LAZY_FP_SHARING)) ||	\
```
why this change?

xodus7 replied:
I believe this was done by the uncrustify tool. There are a few other changes that crept in due to me running it on the files I changed.

ioannisg commented on tests/kernel/fp_sharing/float_disable/src/k_float_disable.c:
```diff
@@ -14,7 +14,7 @@
  */
```
You need to enable the test in .yaml for RISCV

xodus7 replied:
I pushed a commit enabling these tests, however there are still a few missing pieces. PR #23265 adds the options to set the -mabi and -march flags. And I'm still waiting for my QEMU patch to be merged that will allow the qemu_riscv32 & qemu_riscv64 platforms to be FP enabled (or new board targets). I've added filters to the yaml for RISC-V so only targets with CONFIG_CPU_HAS_FPU will be included. I don't believe there are any to date...

ioannisg reviewed (dismissed):
Thanks for this contribution!!

Please, do the following
- enable the tests for RISC-v ARCH and/or platforms, in the respective .yaml files
- extend the documentation  of kernel FP services, to include the support for RISC-V

Commit:
tests: fp_sharing: Enable tests for RISC-V architectures

Enable generic and float_disable tests for RISC-V targets that have
CONFIG_CPU_HAS_FPU enabled.

Signed-off-by: Corey Wharton <coreyw7@fb.com>

Commit:
riscv: Set mabi and march flags for floating point

Adds handling of the FLOAT_64BIT option when determining the ISA
flags as well as introduces a new Kconfig option to enable/disable
the hard-float calling convention.

Signed-off-by: Corey Wharton <coreyw7@fb.com>

xodus7 commented:
@ioannisg @andrewboie @kgugala @carlescufi Any more feedback on this PR?

carlescufi commented:
@ioannisg or @andrewboie any comments before we merge this?

ioannisg commented:
> @ioannisg or @andrewboie any comments before we merge this?

@carlescufi yes, I'd like the test-rework in #23894  to come in first - then this PR should be rebased.

ioannisg reviewed (changes requested):
Need the reviewed test-rework to come in first, then this PR should follow 
after rebase and indication that the test is passing

carlescufi commented:
@xodus7 would you mind rebasing now that #23894 is merged?

xodus7 commented:
> @xodus7 would you mind rebasing now that #23894 is merged?

Done!

stephanosio commented on arch/Kconfig:
```diff
@@ -531,6 +531,13 @@ config FP_SHARING
 	  This option allows multiple threads to use the floating point
 	  registers.
 
+config FLOAT_64BIT
+	bool "Double precision floating point"
+	depends on FLOAT
+	depends on RISCV
+	help
+	  This option means that floating point registers are 64bit width.
```
In ARM, this is controlled by the `CPU_HAS_FPU_DOUBLE_PRECISION` symbol:
https://github.com/zephyrproject-https://github.com/zephyrproject-rtos/zephyr/blob/6492dcec6fa159aa9fb794d69c6abf8fc2b68203/soc/arm/Kconfig#L60

I wonder if we should move that symbol out of `soc/arm` and use that here instead.

ioannisg replied:
by the way, maybe better to write "64bit-wide"

otherwise, yes, that is what i think as well - we should try to unify the Floating-point configuration across architectures

@xodus7 @stephanosio dou you think we could define CPU_HAS_FPU_DOUBLE_PRECISION right under CPU_HAS_FPU definition?

stephanosio replied:
> dou you think we could define CPU_HAS_FPU_DOUBLE_PRECISION right under CPU_HAS_FPU definition?

It sounds like that would be the best place to add it.

stephanosio commented on arch/riscv/Kconfig:
```diff
@@ -13,6 +13,14 @@ config COMPRESSED_ISA
 	bool
 	default y if 64BIT
 
+config FLOAT_HARD
+	bool "Enable hard-float calling convention"
+	default y
+	depends on FLOAT
+	select COMPRESSED_ISA
+	help
+	  This option enables the hard-float calling convention.
```
As far as I can see, there is no ARM `softfp` ABI (i.e. using hardware FP instructions and registers with software floating-point ABI) equivalent in RISC-V.

I wonder if this symbol is really necessary since, as far as the current definition of the `FLOAT` symbol goes, it means hardware FP support is used ([it should really be called `CONFIG_FPU`](https://github.com/zephyrproject-rtos/zephyr/pull/23948#issuecomment-606565206)), in which case this symbol is always implied.

xodus7 replied:
There is a soft-floating point calling convention in RISC-V ([See section 18.3 of this document](https://riscv.org/wp-content/uploads/2015/01/riscv-calling.pdf)) and GCC does support enabling the F or D extensions while using the soft-float calling convention.

stephanosio replied:
After further investigation, I can see that there is indeed ARM `-mfloat-abi=softfp` equivalent in RISC-V:

> `-march=rv32imafdc -mabi=ilp32`: Hardware floating-point instructions can be generated, but no floating-point arguments will be passed in registers. This is like the `-mfloat-abi=softfp` argument to ARM’s GCC, and is usually used when interfacing with soft-float binaries on a hard-float system.

_(quoted from https://gnu-mcu-eclipse.github.io/toolchain/riscv/)_

ioannisg reviewed (commented):
The PR looks good to me - there is a single issue to be resolved, with respect to double-precision FP support, see relevant comment :)

cwshu commented on tests/kernel/fp_sharing/float_disable/src/k_float_disable.c:
```diff
@@ -78,7 +78,7 @@ void test_k_float_disable_common(void)
 		usr_fp_thread.base.user_options);
 
-#if defined(CONFIG_ARM)
+#if defined(CONFIG_ARM) || defined(RISCV)
```
Find a typo. It should be `CONFIG_RISCV`.

Commit:
kconfig: Make the CPU_HAS_FPU_DOUBLE_PRECISION option global.

This option now applies to the RISC-V architecture and is no longer
a ARM only configuration.

Signed-off-by: Corey Wharton <coreyw7@fb.com>

xodus7 commented:
@ioannisg I made the requested change regarding the double precision configuration. However, I've discovered a critical bug while testing on FPGA hardware not caught in the tests running QEMU. The fix is simple and I'm preparing a commit now, but please hold off on merging this for now.

Commit:
riscv: Restore floating-point caller saved registers before integer

This fixes an issue where the t0 register is overwritten after it
has been restored.

Signed-off-by: Corey Wharton <coreyw7@fb.com>

xodus7 commented:
@ioannisg I've pushed a commit that fixes the aforementioned bug.

stephanosio reviewed (approved):
The configs and implementation in general look ok to me (haven't looked deep into the arch details, as I am not an RISC-V expert).

ioannisg commented:
> @ioannisg I've pushed a commit that fixes the aforementioned bug.

This looks fine. I guess it is either t0 or t2 the register that needs to be preserved after being restored (I saw the same fix was applied in 2 cases in the code)

ioannisg commented:
@kgugala want to take a look on this as well?

ioannisg commented:
>This patch was fully tested under the qemu_riscv32 (hard and soft float, 32-bit & 64-bit FLEN) and qemu_riscv64 (hard-float, 64-bit FLEN). However, I had to create a patch in order to enable F & D extensions for the sifive_e machine. I'll be submitting this patch to the QEMU project shortly, but that means I cannot enable emulated testing for this change just yet. Once the required support to QEMU is added I can submit my patch to enabled the required tests.

@xodus7 could you please give us an update on when we expect support in QEMU for this? This seems important because, right now there is no emulated testing in CI for RiscV FP, even if the tests are enabled for risc-v architecture.

In fact, what worries me a little bit more, is that there seems to be no riscv SoC with selected FPU capabilities (CPU_HAS_FPU), so the submitted patch is not even built for any platform (if I am wrong, pls, correct me).

I've got no objections to the patch (great work actually) but I'd like to see a path forward with respect to enabling testing of this feature in the Zephyr.

xodus7 commented:
> @xodus7 could you please give us an update on when we expect support in QEMU for this? This seems important because, right now there is no emulated testing in CI for RiscV FP, even if the tests are enabled for risc-v architecture.
>

I have a patch submitted to the QEMU project that adds a E34 CPU type and allows the sifive_e machine's CPU type to be changed: https://patchew.org/QEMU/20200313193429.8035-1-coreyw7@fb.com/ . It's been reviewed and there haven't been any objections, however, this is my first time submitting to the QEMU project so I'm not sure on the timeline for when it will be merged or when that change will make it into the Zephyr toolchain.

I also have a commit prepared that will select the appropriate CPU based on CONFIG_FLOAT for the qemu_riscv32 and qemu_riscv64 boards. And this will enable the kernel/fp_sharing tests for those platforms.
 
> In fact, what worries me a little bit more, is that there seems to be no riscv SoC with selected FPU capabilities (CPU_HAS_FPU), so the code submitted in this patch is not even built for any Zephyr platform (if I am wrong, pls, correct me).
> 
> I've got no objections to the patch (great work actually) but I'd like to see a path forward with respect to enabling coverage of this feature in the Zephyr. 



You are correct there are no targets currently building this. The only interim solution I can think of is to enable FPU support for the QEMU boards and have them build but not run the kernel/fp_sharing tests. I don't know if that is supported (build-only) or if it would be an acceptable temporary solution?

ioannisg commented:
I guess that if you can enable the fpu on the riscv qemu it would be great. I believe that there is an option for build only in Zephyr tests, see test yaml files for info :)

Commit:
tests: fp_sharing: Enable build-only tests for RISC-V QEMU platform

Enables build-only tests for the qemu_riscv32 platform.

Signed-off-by: Corey Wharton <coreyw7@fb.com>

xodus7 commented:
@ioannisg I've enabled the FPU for the qemu_riscv32 board and made the corresponding tests build-only. Unfortunately, I couldn't do the same for the qemu_riscv64 board because it was generating an illegal instruction fault when running tests/lib/sprintf/libraries.libc.sprintf; specifically the in the test_sprintf_double() function. Does this provide sufficient coverage for now?

ioannisg commented:
> ing tests/lib/sprintf/libraries.libc.sprintf

@xodus7 I think the issue with that test is that it sets CONFIG_FLOAT=y, so effectively, it is doing something similar to the dedicated kernel FP tests :)

AFAICS, that is the only test of this kind, which does not filter-out riscv architecture in CI.
I've checked out your branch and ran locally the test for riscv32 - it runs OK. I can also reproduce the failure for riscv64 in qemu.

1. Do you suspect this is a QEMU bug?
2. Can you file an bug report in Zephyr for this specific failure?

I don't have a good proposal on how to handle this: you could enable FPU in qemu_riscv64 and  skip the `test_sprintf_double `test from running but this is not ideal. Or, leave the PR as is, but then there's no build coverage for FPU in RISCV64.

Btw, I think you should enable the FPU at the SoC level, in your case, that should be the SiFive Freedom, correct?

ioannisg reviewed (approved):
I'd like to unblock this work so it can be merged. I think my last comments should be addressed and also, I'd like risc-v maintainers to have another look, if possible.

xodus7 commented:
> AFAICS, that is the only test of this kind, which does not filter-out riscv architecture in CI.
> I've checked out your branch and ran locally the test for riscv32 - it runs OK. I can also reproduce the failure for riscv64 in qemu.
> 
> 1. Do you suspect this is a QEMU bug?

No, the qemu_riscv32 board has only single-precision enabled, so no actual FPU instructions are generated in this test because it operates on double. Additionally, CONFIG_FP_SHARING is not enabled nor do any threads have K_FP_REGS set so no floating-point context is saved.

> 2. Can you file an bug report in Zephyr for this specific failure?

I don't think there is a bug here. Am I wrong?

> 
> I don't have a good proposal on how to handle this: you could enable FPU in qemu_riscv64 and skip the `test_sprintf_double `test from running but this is not ideal. Or, leave the PR as is, but then there's no build coverage for FPU in RISCV64.
> 

I think there is little risk here since the double-precision code is almost identical to the single-precision. Since this is only an interim solution I think the right thing to do is leave it the way it is. Thoughts?

> Btw, I think you should enable the FPU at the SoC level, in your case, that should be the SiFive Freedom, correct?
> 
> **Edit**: I saw, now, that both qemu_riscv targets are based on the same SoC so maybe makes more sense to keep enabling FPU on the board level.

Well, the SiFive Freedom E310 is based on a E31 and has no FPU so we don't want to do that; however, it is the basis for the machine we are using for QEMU emulation. Enabling at the board level is the easiest way to add FPU support to RISCV emulation (when QEMU is patched) without creating a new QEMU board and SOC.

galak commented on soc/arm/Kconfig:
```diff
@@ -57,14 +57,6 @@ config NRF_SPU_RAM_REGION_SIZE
 
-config CPU_HAS_FPU_DOUBLE_PRECISION
-	bool
-	depends on CPU_CORTEX_M7
```
Do we need to be selecting this on CPU_CORTEX_M7 now?

ioannisg replied:
Not really, @galak , this was more of an extra precaution not to select the option outside Cortex-M7. But I think we can live without it.

stephanosio replied:
All the relevant Cortex-M SoCs (i.e. the ones configured with DP support) already select this symbol.

As for `depends on`, I agree with @ioannisg.

galak commented:
> The relevant QEMU patchset: http://patchwork.ozlabs.org/project/qemu-devel/list/?series=164096

Seems like reasonable patches to pull into the zephyr SDK

xodus7 commented:
> > The relevant QEMU patchset: http://patchwork.ozlabs.org/project/qemu-devel/list/?series=164096
> 
> Seems like reasonable patches to pull into the zephyr SDK

@galak I wasn't aware of this, thanks! I created a PR zephyrproject-rtos/sdk-ng#204 with these patches.

carlescufi commented:
@xodus7 could you please rebase?
@galak should we wait until https://github.com/zephyrproject-rtos/sdk-ng/pull/204 is merged to merge this PR?

carlescufi commented:
@xodus7 did you make a mistake while rebasing? a lot of unrelated commits show here now.

xodus7 commented:
I guess I did! I'm not sure what I did wrong....

carlescufi commented:
> I guess I did! I'm not sure what I did wrong....

well, not sure myself but you'd typically do `git rebase upstream/master` to rebase a PR

xodus7 commented:
So I did it again and this time remembered to push master to my forked repo and that seems to have resolved it. Sorry about the noise!

andrewboie commented on arch/Kconfig:
```diff
@@ -516,7 +523,7 @@ config MPU_GAP_FILLING
 	bool "Floating point"
 	depends on CPU_HAS_FPU
-	depends on ARM || X86 || ARC
+	depends on ARC || ARM || RISCV || X86
```
passing comment: really should just delete this line and rely solely on CPU_HAS_FPU which should be selected by CPUs that have this capability

tyrantlizard77 reviewed (commented):
misc whining about Kconfig (which doesn't really need to be addressed in this PR), question about 2 spots where interrupts are locked but might be better to just lock the scheduler instead and still allow IRQs

xodus7 commented on arch/riscv/core/thread.c:
```diff
@@ -58,7 +59,78 @@ void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
+	}
+
+	/* Ensure a preemptive context switch does not occur */
+	key = irq_lock();
```
I think we really don't want any context switch including IRQs to fire here. I haven't tested the case where an interrupt fires here before the work is done and I'd be surprised if there wasn't some corner case not well handled. Such as restoring FP registers that haven't been saved yet.

xodus7 commented on arch/Kconfig:
```diff
@@ -464,6 +464,13 @@ config CPU_HAS_FPU
 	  This option is enabled when the CPU has hardware floating point
 	  unit.
 
+config CPU_HAS_FPU_DOUBLE_PRECISION
```
This used to be an ARM only option for Cortex-M7 processors. :-)
